### ISSUE TITLE
EnergyCalibration: energy2ph fails with use_approximation and use_zerozero

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#57]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/57) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

```
In [59]: cal2._use_loglog, cal2._use_zerozero, cal2._use_approximation
Out[59]: (False, True, True)

In [60]: cal2.energy2ph(5000)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-60-2406f84365a8> in <module>()
----> 1 cal2.energy2ph(5000)

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mass-0.4.3-py2.7-macosx-10.10-x86_64.egg/mass/calibration/energy_calibration.pyc in energy2ph(self, energy)
    158             self._update_converters()
    159         energy_residual = lambda ph, etarget: self._ph2energy_anon(ph)-etarget
--> 160         return sp.optimize.brentq(energy_residual, 1e-6, self._max_ph, args=(energy,))
    161 
    162     def energy2dedph(self, energy, denergy=1):

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/scipy/optimize/zeros.pyc in brentq(f, a, b, args, xtol, rtol, maxiter, full_output, disp)
    436     if rtol < _rtol:
    437         raise ValueError("rtol too small (%g < %g)" % (rtol, _rtol))
--> 438     r = _zeros._brentq(f,a,b,xtol,rtol,maxiter,args,full_output,disp)
    439     return results_c(full_output, r)
    440 

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mass-0.4.3-py2.7-macosx-10.10-x86_64.egg/mass/calibration/energy_calibration.pyc in <lambda>(ph, etarget)
    157         if self._model_is_stale:
    158             self._update_converters()
--> 159         energy_residual = lambda ph, etarget: self._ph2energy_anon(ph)-etarget
    160         return sp.optimize.brentq(energy_residual, 1e-6, self._max_ph, args=(energy,))
    161 

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mass-0.4.3-py2.7-macosx-10.10-x86_64.egg/mass/mathstat/interpolate.pyc in __call__(self, x, der)
    377             splresult[low] = self.lowline(x[low]-self.x[0])
    378         if np.any(high):
--> 379             splresult[high] = self.highline(x[high]-self.x[-1])
    380         return splresult
    381 

TypeError: 'float' object has no attribute '__getitem__'
> /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mass-0.4.3-py2.7-macosx-10.10-x86_64.egg/mass/mathstat/interpolate.py(379)__call__()
    377             splresult[low] = self.lowline(x[low]-self.x[0])
    378         if np.any(high):
--> 379             splresult[high] = self.highline(x[high]-self.x[-1])
    380         return splresult
    381 
```

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Actually it crahses for both use_zerozero True and false, as long as use_loglog is False

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

note that in the most recent develop there is a file "test_calibration.py", all the commented out tests fail because of this issue

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Problems come from mass.mathstat.interpolate objects SmoothingSpline and CubicSpline.

They do not (yet) have consistent behavior for the form of their output depending on whether the input is an array, a list, or a scalar. Working on getting it there.

BTW, this is and always has been a serious hassle about writing functions that act successfully on numpy arrays or scalars. I assume the behavior we want is scalar -> scalar and sequence (array, tuple, list, whatever) -> array. Various questions on stack exchange lead me to believe that there's not one single, canonical and simple way to achieve this.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Make EnergyCalibration.ph2energy(x) and .energy2ph(x) return values consistent.

Return values should match x as scalars/arrays. That is, scalar in = scalar out.
I had to change 0-d arrays to actual scalars by x[()]. Ugh.

Also, I added feature to EnergyCalibration.energy2ph(x) so that it will accept
a non-scalar input for x. Dangerous! That could be slow for long x. Fixes #57.
